device/src/secure_boot.rs

```diff
@@ -0,0 +1,398 @@
+    rsa_public_modulus: [u8; 384],
+    rsa_public_exponent: [u8; 4],
+    precalculated_r: [u8; 384],
+    precalculated_m_prime: [u8; 4],
+    rsa_pss_signature: [u8; 384],
+    crc32: [u8; 4],
+}
+
+impl SignatureBlock {
+    fn from_bytes(data: &[u8]) -> Self {
+        let mut block = SignatureBlock {
+            image_digest: [0; 32],
+            rsa_public_modulus: [0; 384],
+            rsa_public_exponent: [0; 4],
+            precalculated_r: [0; 384],
+            precalculated_m_prime: [0; 4],
+            rsa_pss_signature: [0; 384],
+            crc32: [0; 4],
+        };
+
+        block.image_digest.copy_from_slice(&data[4..36]);
+        block.rsa_public_modulus.copy_from_slice(&data[36..420]);
+        block.rsa_public_exponent.copy_from_slice(&data[420..424]);
+        block.precalculated_r.copy_from_slice(&data[424..808]);
+        block.precalculated_m_prime.copy_from_slice(&data[808..812]);
+        block.rsa_pss_signature.copy_from_slice(&data[812..1196]);
+        block.crc32.copy_from_slice(&data[1196..1200]);
+
+        block
+    }
+}
+
+// Convert bytes to u32 array in little-endian format (ESP32-C3 native format)
+fn bytes_to_u32_le_native(bytes: &[u8]) -> Vec<u32> {
+    bytes
+        .chunks(4)
+        .map(|chunk| {
+            let mut word = [0u8; 4];
+            word[..chunk.len()].copy_from_slice(chunk);
+            u32::from_le_bytes(word)
+        })
+        .collect()
+}
+
+fn verify_rsa_pss_signature(
+    rsa: &mut Rsa<'_, Blocking>,
+    sig_block: &SignatureBlock,
+    message_hash: &[u8; 32],
+    sha: &mut Sha,
+) -> Result<bool, &'static str> {
+    const KEY_SIZE_BYTES: usize = 384;
+
+    // Wait for RSA peripheral to be ready
+    block!(rsa.ready()).map_err(|_| "RSA peripheral not ready")?;
+
+    // Convert signature block data to u32 arrays in ESP32-C3 native little-endian format
+    // All inputs should be little-endian byte arrays
+    let modulus_u32 = bytes_to_u32_le_native(&sig_block.rsa_public_modulus);
+    let exponent_u32 = bytes_to_u32_le_native(&sig_block.rsa_public_exponent);
+    let r_u32 = bytes_to_u32_le_native(&sig_block.precalculated_r);
+    // No byte reversal for signature - use native little-endian format
+    let signature_u32 = bytes_to_u32_le_native(&sig_block.rsa_pss_signature);
+    let m_prime = u32::from_le_bytes(sig_block.precalculated_m_prime);
+
+    // Convert to fixed-size arrays for Op3072 (96 u32 words = 384 bytes = 3072 bits)
+    if modulus_u32.len() != 96 || r_u32.len() != 96 || signature_u32.len() != 96 {
+        return Err("Invalid RSA key size - expected 3072 bits");
+    }
+
+    let mut modulus = [0u32; 96];
+    let mut r_value = [0u32; 96];
+    let mut signature = [0u32; 96];
+    let mut exponent = [0u32; 96];
+
+    modulus.copy_from_slice(&modulus_u32);
+    r_value.copy_from_slice(&r_u32);
+    signature.copy_from_slice(&signature_u32);
+    exponent[0] = exponent_u32[0]; // Exponent is typically small, just copy first word
```

Comment:
Typically or always for this variant of RSA PSS? Is there a reference for this? Any downside to of copying it all? Probably doesn't matter given it works